(WIP) Major v16 tutorial edits

a-velasco opened the pull request:
## Issue
The tutorial for PostgreSQL 16 has outputs and instructions that are out of date.

## Solution
Test and rewrite the tutorial. Main changes:
* General improvements in clarity, wording, formatting
* Updated outputs for all commands

Additional updates:
* Renamed all mentions of `16/edge` to `16/stable`. 

## Checklist
- [x] I have added or updated any relevant documentation.
- [x] I have cleaned any remaining cloud resources from my accounts.

taurus-forever commented on docs/tutorial/index.md:
Proposal: let's add to every tutorial (and other manuals?):
a) `Course Levels 101` for Tutorial https://course-faq.com/what-do-the-numbers-in-course-names-mean
b) `Duration: 25:00`  for Tutorial (see [MAAS example](https://maas.io/tutorials/build-your-own-bare-metal-cloud-using-a-raspberry-pi-cluster-with-maas#1-overview), so users can properly prepare for tutorial).

a-velasco replied:
a) I think introducing a standard like this has a few conflicts with Diataxis and our general documentation practices. I understand what you want to do though, and I have several thoughts about other avenues.
b) Agree with adding approximate duration.

taurus-forever commented on docs/tutorial/index.md:
```diff
@@ -317,148 +385,112 @@ Removing a unit from the application scales down the replicas.
+Machine  State    Address                                 Inst id        Base          AZ  Message
+0        started  10.26.224.154                           juju-1c143d-0  ubuntu@24.04      Running
+1        started  10.26.224.142                         
+  juju-1c143d-1  ubuntu@24.04      Running
```
Layout issue.

taurus-forever commented on docs/tutorial/index.md:
```diff
@@ -317,148 +385,112 @@ Removing a unit from the application scales down the replicas.
+Machine  State    Address                                 Inst id        Base          AZ  Message
+0        started  10.26.224.154                           juju-1c143d-0  ubuntu@24.04      Running
+1        started  10.26.224.142                         
+  juju-1c143d-1  ubuntu@24.04      Running
```
layout issue

taurus-forever commented on docs/tutorial/index.md:
```diff
@@ -317,148 +385,112 @@ Removing a unit from the application scales down the replicas.
+Machine  State    Address                                 Inst id        Base          AZ  Message
+0        started  10.26.224.154                           juju-1c143d-0  ubuntu@24.04      Running
+1        started  10.26.224.142                         
+  juju-1c143d-1  ubuntu@24.04      Running
```
layout issue.

taurus-forever commented on docs/tutorial/index.md:
```diff
@@ -570,76 +608,105 @@ TLS is enabled by integrating Charmed PostgreSQL with the [Self-signed certifica
+Machine  State    Address                                 Inst id        Base          AZ  Message
+0        started  10.26.224.154                           juju-1c143d-0  ubuntu@24.04      Running
+1        started  10.26.224.142                         
+  juju-1c143d-1  ubuntu@24.04      Running
```
layout

taurus-forever commented on docs/tutorial/index.md:
````diff
@@ -105,135 +145,151 @@ You can also watch juju logs with the [`juju debug-log`](https://juju.is/docs/ju
 ```text
-postgres=## CREATE DATABASE mynewdatabase;
-postgres=## \c mynewdatabase
+postgres=# CREATE DATABASE mynewdatabase;
````
We should mentione here something like: we are creating database here manually for edication purpouse only, while the proper way has been described below using data-integrator or direct relation with db client app.

As we are not recomending used to create Databases manually (objects ownership, etc).

taurus-forever commented on docs/tutorial/index.md:
```diff
@@ -282,32 +342,40 @@ To enable high availability in a production environment, replicas should be host
-postgresql/0*  active    idle   0        10.89.49.129           Primary
-postgresql/1   active    idle   1        10.89.49.197           
-postgresql/2   active    idle   2        10.89.49.175           
+Unit           Workload  Agent  Machine  Public address                          Ports     Message
+postgresql/0*  active    idle   0        10.26.224.154                           5432/tcp  Primary
+postgresql/1   active    idle   1        10.26.224.142                         
+  5432/tcp
+postgresql/2   active    idle   2        10.26.224.123                           5432/tcp
```
Here some problems with layout, between columns "Public address" and "Ports"

taurus-forever commented on docs/tutorial/index.md:
```diff
@@ -282,32 +342,40 @@ To enable high availability in a production environment, replicas should be host
-0        started  10.89.49.129  juju-a8a31d-0  jammy       Running
-1        started  10.89.49.197  juju-a8a31d-1  jammy       Running
-2        started  10.89.49.175  juju-a8a31d-2  jammy       Running
+Machine  State    Address                                 Inst id        Base          AZ  Message
+0        started  10.26.224.154                           juju-1c143d-0  ubuntu@24.04      Running
+1        started  10.26.224.142                         
+  juju-1c143d-1  ubuntu@24.04      Running
+2        started  10.26.224.123                           juju-1c143d-2  ubuntu@24.04      Running
```
Extra line 377 and columns layout issue (too wide between Address and Inst id)

taurus-forever reviewed (commented):
Nice update!

dragomirp commented on docs/explanation/units.md:
```diff
@@ -26,7 +26,7 @@ Model       Controller  Cloud/Region         Version  SLA          Timestamp
 
 App         Version  Status  Scale  Charm       Channel    Rev  Exposed  Message
-postgresql  14.15    active      3  postgresql  14/stable  553  no       
+postgresql  16.19    active      3  postgresql  16/stable  843  no       
```
Current version is 16.9.

dragomirp commented on docs/how-to/manage-passwords.md:
````diff
@@ -29,7 +33,11 @@ juju grant-secret <secret_name> postgresql
+```
+
+```{tip}
+Note that `<secret_URI>` includes the `secret:` prefix.
````
IIRC the prefix is called scheme in URIs.

dragomirp commented on docs/tutorial/index.md:
````diff
@@ -105,135 +145,155 @@ You can also watch juju logs with the [`juju debug-log`](https://juju.is/docs/ju
-```text
-juju run postgresql/leader get-password
+```{seealso}
+For more information about password management with Juju secrets, see [](/how-to/manage-passwords)
````
Should the `[]` be empty here?

a-velasco replied:
Yes, it auto-fills with the title of the page :) Fancy MyST feature so we don't have to worry about pages getting renamed.

dragomirp commented on docs/tutorial/index.md:
```diff
@@ -105,135 +145,155 @@ You can also watch juju logs with the [`juju debug-log`](https://juju.is/docs/ju
-While still in the leader unit's shell, run the command below to list all databases currently available:
+We'll need the IP address associated with the specific application unit we want to interact with. You can find it with `juju status`. 
+
+Since we will use the leader unit to connect to PostgreSQL, we are interested in the address for the unit marked with `*`, like in the output below:
```
We should most probably instruct users to look for the primary, not the juju leader. It shouldn't matter for the tutorial, but in principle the leader and thee primary could be different.

dragomirp commented on docs/tutorial/index.md:
```diff
@@ -282,32 +346,38 @@ To enable high availability in a production environment, replicas should be host
 ### Add units
 
-Currently, your deployment has only one juju **unit**, known in juju as the **leader unit**. You can think of this as the database **primary instance**. For each **replica**, a new unit is created. All units are members of the same database cluster.
+Currently, your deployment has only one Juju unit, known in juju as the **leader unit**. You can think of this as the database's **primary instance**. For each cluster replica, a new Juju unit is created. 
```
This is somewhat misleading, since Juju and Patroni track leadership separately. When deploying, they will match, but may drift apart over time.

dragomirp commented on docs/tutorial/index.md:
````diff
@@ -317,245 +387,239 @@ Removing a unit from the application scales down the replicas.
 
-Running the command should output something like:
+```{terminal}
+:input: juju deploy data-integrator --config database-name=mydb-integrator
````
There's now a noble based data-int in stable that can be deployed with `--base ubuntu@24.04`.

a-velasco commented:
Merging due to urgency of these updates (banner removal, updated tutorial, updated status outputs). Items left to address:
* `data-integrator` deployment command in tutorial
* Better way of explaining leader unit vs. primary
* Failed doc checks (fixed when https://github.com/canonical/postgresql-operator/pull/1062 gets merged)

a-velasco commented on docs/how-to/enable-ldap.md:
```diff
@@ -30,7 +30,7 @@ Deploy the [GLAuth charm](https://charmhub.io/glauth-k8s):
 juju add-model glauth
 juju deploy self-signed-certificates
-juju deploy postgresql-k8s --channel 16/stable --trust
+juju deploy postgresql-k8s --channel 14/stable --trust
```
Resolved offline. TLDR: `14/stable` for this particular instance of `postgresql-k8s` is correct; we'll update it when `16/stable` is out.
